server/src/main/java/org/cloudfoundry/identity/uaa/zone/MultitenantVmidentityClientDetailsService.java

```diff
@@ -0,0 +1,335 @@
+/*******************************************************************************
+ *     Cloud Foundry
+ *     Copyright (c) [2017] VMware, Inc. All Rights Reserved.
+ *
+ *     This product is licensed to you under the Apache License, Version 2.0 (the "License").
+ *     You may not use this product except in compliance with the License.
+ *
+ *     This product includes a number of subcomponents with
+ *     separate copyright notices and license terms. Your use of these
+ *     subcomponents is subject to the terms and conditions of the
+ *     subcomponent's license, as noted in the LICENSE file.
+ *******************************************************************************/
+package org.cloudfoundry.identity.uaa.zone;
+
+import java.security.NoSuchAlgorithmException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.cloudfoundry.identity.uaa.client.MultitenantClientDetailsService;
+import org.cloudfoundry.identity.uaa.client.VmidentityDataAccessException;
+import org.cloudfoundry.identity.uaa.oauth.client.ClientConstants;
+import org.cloudfoundry.identity.uaa.util.VmidentityUtils;
+import org.springframework.security.core.GrantedAuthority;
+import org.springframework.security.core.authority.AuthorityUtils;
+import org.springframework.security.crypto.password.PasswordEncoder;
+import org.springframework.security.oauth2.provider.ClientAlreadyExistsException;
+import org.springframework.security.oauth2.provider.ClientDetails;
+import org.springframework.security.oauth2.provider.ClientRegistrationException;
+import org.springframework.security.oauth2.provider.NoSuchClientException;
+import org.springframework.security.oauth2.provider.client.BaseClientDetails;
+import org.springframework.util.Assert;
+
+import com.google.common.primitives.Ints;
+import com.vmware.identity.idm.DuplicatedOIDCClientException;
+import com.vmware.identity.idm.DuplicatedOIDCRedirectURLException;
+import com.vmware.identity.idm.NoSuchOIDCClientException;
+import com.vmware.identity.idm.OIDCClient;
+import com.vmware.identity.idm.client.CasIdmClient;
+
+public class MultitenantVmidentityClientDetailsService implements MultitenantClientDetailsService {
+
+    private static final Log logger = LogFactory.getLog(MultitenantVmidentityClientDetailsService.class);
+
+    private final CasIdmClient idmClient;
+    private final PasswordEncoder passwordEncoder;
+
+    public MultitenantVmidentityClientDetailsService(CasIdmClient idmClient, PasswordEncoder passwordEncoder) throws NoSuchAlgorithmException {
+        Assert.notNull(idmClient, "CasIdmClient required");
+        this.idmClient = idmClient;
+        this.passwordEncoder = passwordEncoder;
```

Comment:
We can, but I don't think it's a major concern. The JDBC variant doesn't assert anything when setting the encoder and it's all wired together with the Spring configs.